pkg/pgmodel/upgrade_tests/upgrade_test.go

```diff
@@ -229,78 +245,123 @@ func withDBAndConnector(t testing.TB, DBName string, connectorImage string,
 		log.Fatal(err)
 	}
 
-	dbContainer, err := testhelpers.StartPGContainerWithImage(ctx, cleanImage, tmpDir, *printLogs)
+	dataDir, err := testhelpers.TempDir("update_test_data")
 	if err != nil {
-		t.Fatal("Error setting up container", err)
+		log.Fatal(err)
 	}
 
-	defer testhelpers.StopContainer(ctx, dbContainer, *printLogs)
+	// Start a db with the prev extension and a prev connector as well
+	// Then run preUpgrade and shut everything down.
+	func() {
+		fmt.Println("PrevImage", prevDBImage)
+		dbContainer, err := testhelpers.StartPGContainerWithImage(ctx, prevDBImage, tmpDir, dataDir, *printLogs, *useExtension)
+		if err != nil {
+			t.Fatal("Error setting up container", err)
+		}
 
-	db, err := testhelpers.DbSetup(*testDatabase, testhelpers.Superuser)
-	if err != nil {
-		t.Fatal(err)
-		return
-	}
-	db.Close()
+		defer testhelpers.StopContainer(ctx, dbContainer, *printLogs)
 
-	connector, err := testhelpers.StartConnectorWithImage(context.Background(), connectorImage, *printLogs, *testDatabase)
-	if err != nil {
-		log.Fatal(err.Error())
-	}
-	defer testhelpers.StopContainer(ctx, connector, *printLogs)
+		db, err := testhelpers.DbSetup(*testDatabase, testhelpers.Superuser)
+		if err != nil {
+			t.Fatal(err)
+			return
+		}
+		db.Close()
 
-	connectorHost, err := connector.Host(ctx)
-	if err != nil {
-		t.Fatal(err)
-		return
-	}
+		connectorImage := "timescale/timescale-prometheus:" + prevVersion.String()
+		connector, err := testhelpers.StartConnectorWithImage(context.Background(), connectorImage, *printLogs, *testDatabase)
+		if err != nil {
+			log.Fatal(err.Error())
+		}
+		defer testhelpers.StopContainer(ctx, connector, *printLogs)
+
+		connectorHost, err := connector.Host(ctx)
+		if err != nil {
+			t.Fatal(err)
+			return
+		}
 
-	connectorPort, err := connector.MappedPort(ctx, testhelpers.ConnectorPort)
+		connectorPort, err := connector.MappedPort(ctx, testhelpers.ConnectorPort)
+		if err != nil {
+			t.Fatal(err)
+			return
+		}
+		preUpgrade(dbContainer, tmpDir, connectorHost, connectorPort)
+	}()
+
+	//Start a new connector and migrate.
+	//Then run postUpgrade
+	dbContainer, err := testhelpers.StartPGContainerWithImage(ctx, cleanImage, tmpDir, dataDir, *printLogs, *useExtension)
 	if err != nil {
-		t.Fatal(err)
-		return
+		t.Fatal("Error setting up container", err)
 	}
-	f(dbContainer, tmpDir, connectorHost, connectorPort)
+
+	defer testhelpers.StopContainer(ctx, dbContainer, *printLogs)
+
+	t.Logf("upgrading versions %v => %v", prevVersion, version.Version)
+	// TODO this should be NoSuperuser expect for the extwlist error
+	connectURL := testhelpers.PgConnectURL(*testDatabase, testhelpers.Superuser)
+	migrateToVersion(t, connectURL, version.Version, "azxtestcommit")
```

Comment:
(with the need to reset the pristine DB as well, performing the migrate here may be forced by the reset logic, so this comment may be irrelevant. Still, if we do end up having the discussion I think it's better that we do so explicitly in a well-recorded location instead of just in-person or on slack, so I'm leaving it for now)

so here's the thing: these lines are the most important lines of this file; they define how we're doing the upgrade, and thus the validity of the tests. That being the case, I think they should be high in the call-hierarchy as possible, as any issue the tests encounter will almost certainly relate to them, and this is going to be the place we debug from. As such I think putting them inside this function is a false economy giving the illusion encapsulation for no real benefit.